doc/rst/technotes/gpu.rst

```diff
@@ -97,10 +86,106 @@ kernels for the ``forall`` loops in the function.
+
+* We currently only target NVIDIA GPUs (although we intend to support AMD
+  GPUs in a future release).
+
+* ``LLVM`` must be used as Chapel's backend compiler (i.e.
+  ``CHPL_LLVM`` must be set to ``system`` or ``bundled``). For more information
+  about these settings see :ref:`Optional Settings <readme-chplenv>`.
+
+* If using a system install of ``LLVM`` we expect this to be the same
+  version as the bundled version (currently 14). Older versions may
+  work; however, we only make efforts to test GPU support with this version.
+
+* ``CHPL_TASKS=qthreads`` is required for GPU support.
+
+* PGAS style communication is not available within GPU kernels; that is:
+  reading from or writing to a variable that is stored on a different locale
+  from inside a GPU eligible loop (when executing on a GPU) is not supported.
+
+* There is no user-level feature to specify GPU block size on a
+  per-kernel basis. This can be set on a program wide basis at compile-time by
+  passing ``--gpu-block-size=size`` to the compiler or setting it with the
+  ``CHPL_GPU_BLOCK_SIZE`` environment variable.
+
+* There is no user-level feature to allocate or access block shared memory.
+
+* The use of most ``extern`` functions within a GPU eligible loop is not supported
+  (a limited set of functions used by Chapel's runtime library are supported). 
+
+   * Note that various functions with Chapel's standard module call out to such
+     functions and thus is, in turn, is not supported in such loops.
+
+* For loops to be considered eligible for execution on a GPU they
+  must fulfill the requirements discussed in the `Overview`_ section.
+
+GPU Support Features
 --------------------
 
+In the following subsections we discuss particular features or aspects of
+GPU supports that may be of interest.
+
+Diagnostics and Utilities
+~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The :mod:`GPUDiagnostics` module contains functions to help users count and
+track kernel launches.
+
+To count the number of kernel launches that occur in a section of code,
+surround that code with calls to :proc:`~GPUDiagnostics.startGPUDiagnostics`
+and :proc:`~GPUDiagnostics.stopGPUDiagnostics` and then call
+:proc:`~GPUDiagnostics.getGPUDiagnostics`.  If called in a multi-locale
+environment :proc:`~GPUDiagnostics.getGPUDiagnostics` will return an array of
+counts of launches on a per-locale basis.
+
+To get verbose output (indicating the location of each kernel launch) surround
+the code with calls to :proc:`~GPUDiagnostics.startVerboseGPU` and
+:proc:`~GPUDiagnostics.stopVerboseGPU`. This output will directed to
+``stdout``.
+
+The ``GPU`` module contains additional utility functions. One particularly
+useful function is ``GPU.assertOnGpu()``.  This function will conduct a runtime
+assertion that will halt execution when not being performed on a GPU.  If
+:proc:``GPU.assertOnGpu()`` appears as the first line of ``forall`` or
+``foreach`` loop the Chapel compiler will do a compile-time check and produce
```

Comment:
Isn't it technically "at the top level" of any forall loop? i.e. if it's second after some declaration. I thought I remembered seeing iteration over all children of the loop body.

Reply:
Yes you're right. 

And as we talked about before you could even embed it inside of some control flow so long as that control flow gets eliminated by the time we get to the gpuization pass. IOW this would work:

```chpl
foreach i in ... {
  // ... bunch of straightline code here ...
  if(paramEvaluatableFunctionThatReturnsTrue()) { assertOnGpu(); }
}
```

Trying to explain this to end users is a bit tricky though (what magic happens in between compiler passes is an implementation detail and something that's subject to change) so I just said "put at the top" (where it's guaranteed to work) and I think if it works anywhere else that's just kind of a nice bonus.

I suppose one concern I'd have is if some user really did want this to just be a runtime check they could say: "hey, I didn't want to do the compiletime check so I made it the second line and this is still happening".

So I'll add this sentence:

This check might also occur if :proc:``GPU.assertOnGpu()`` is placed elsewhere in the loop depending on what is compile-time analyzable and the presence of control flow.

Reply:
> This check might also occur if :proc:GPU.assertOnGpu() is placed elsewhere in the loop depending on what is compile-time analyzable and the presence of control flow.

Actually I think I'll further simplify that to:

This check might also occur if :proc:GPU.assertOnGpu() is placed elsewhere in the loop depending on the presence of control flow.